---
title: "PowerShell特有の落とし穴！Function内で標準出力すると戻り値が「System.Object[]」になってしまう"
emoji: "🛫"
type: "tech" # tech: 技術記事 / idea: アイデア
topics: ["powershell"]
published: false
---

PowerShellでは「Function内の標準出力が戻り値として呼び出し元に渡される」という**仕様**があります。

要は『 **Function内で標準出力すると意図しない戻り値になるよ** 』ということ。

何を隠そう私も時間が経つと、この仕様を忘れてしまい何度かハマっています……
私自身の備忘もかねて、この仕様の詳しい情報と対処方法を紹介。

## この記事のターゲット

- 主にPowerShell初学者・初級者の方
- 下記のようなケースの原因不明なエラーでお悩みの方
    - 複数のFunctionで連携するPowerShellスクリプトを実行するとエラーが発生
    - コードを見直すが問題がないように見え、なぜエラーが発生しているか不明
    - エラーメッセージをみると、なぜかFunctionの戻り値が想定外のデータ型で返っている

## 「Function内で標準出力すると意図しない戻り値になる」とは

言葉の通りなのですが文章だけではイメージが掴みにくいと思うので、その事象となるコードを記載します。

### 仕様で戻り値が期待値とならないコード

意図しない戻り値にさせてしまっているポイントには「❌マーク」でコメントしています。

```powershell:仕様で戻り値が期待値とならないコードの例
# Function
function New-Folder {
    param (
        [string]$FolderPath
    )
    # 処理結果の進捗を入れる数値型のコード
    [int]$statusCode = 0

    if (Test-Path -Path $FolderPath) {
        # すでにフォルダーがある場合は削除
        try {
            Remove-Item -Path $FolderPath -Recurse -Force
        }
        catch {
            Write-Error "New-Folder: フォルダー削除でエラーが発生。FolderPath[$($FolderPath)]エラー[$($_.Exception.Message)]"
            $statusCode = -101
        }

        # 削除後にフォルダーを作成
        if ($statusCode -eq 0) {
            try {
                # ❌ New-Itemのコマンド結果が戻り値に入る
                New-Item -ItemType Directory -Path $FolderPath
                # ❌ Write-Outputのコマンド結果が戻り値に入る
                Write-Output "New-Folder: フォルダーを再作成。"
            }
            catch {
                Write-Error "New-Folder: フォルダー作成でエラーが発生。FolderPath[$($FolderPath)]エラー[$($_.Exception.Message)]"
                $statusCode = -201
            }
        }
    }
    else {
        # フォルダーがない場合はそのまま作成
        try {
            # ❌ New-Itemのコマンド結果が戻り値に入る
            New-Item -ItemType Directory -Path $FolderPath
            # ❌ Write-Outputのコマンド結果が戻り値に入る
            Write-Output "New-Folder: フォルダーを新規作成。"
        }
        catch {
            Write-Error "New-Folder: フォルダー作成でエラーが発生。FolderPath[$($FolderPath)]エラー[$($_.Exception.Message)]"
            $statusCode = -301
        }
    }

    # ❌ Write-Outputのコマンド結果が戻り値に入る
    if ($statusCode -eq 0) {
        Write-Output "New-Folderが正常終了。statusCode[$($statusCode)]"
    }
    else {
        Write-Output "New-Folderが異常終了。statusCode[$($statusCode)]"
    }

    # ❌ Intで返ったほしいが、標準出力の文字列型も入ってObject型として返してしまう。
    return $statusCode
}

# メイン処理
[int]$statusCode = 0

# ❌ Function実行している場所で実行するとObject型で返ってきてくる。
#     戻り値をInt型の$statusCodeに代入しているのでデータ型の変換処理が発生。Object型からInt型には変換不可となるためココでエラーが発生。
#     $stautsCode に値を代入するタイミングでエラーが発生しているため、$statusCodeは「0」のままとなる。
$statusCode = (New-Folder "D:\Temp")

# ❌ 上記でエラーが発生しているものの$statusCodeは「0」のままなので正常終了のメッセージが出力
if ($statusCode -eq 0) {
    Write-Host "メイン処理が正常終了。statusCode[$($statusCode)]"
}
else {
    Write-Host "メイン処理が異常終了。statusCode[$($statusCode)]"
}
```

このようなコードを実行すると Function（`New-Folder`）の戻り値は、数値型（`Int`）ではなく Function内で出力された標準出力結果（`Write-Output` や `New-Item`）を含む オブジェクト型（`System.Object[]`）が戻り値に。

つぎは実際にこのコードを実行しています。

### “仕様で戻り値が期待値とならないコード”を実行した結果

“仕様で戻り値が期待値とならない”コードを実行すると下記のようなエラーが発生します。

```powershell:エラー（結果をそのまま記載）
MetadataError:
Line |
  63 |  $statusCode = (New-Folder "D:\Temp")
     |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot convert the "System.Object[]" value of type "System.Object[]" to type "System.Int32".
メイン処理が正常終了。statusCode[0]
```

日本語の翻訳すると下記のとおり。

```powershell:エラー（結果を日本語に翻訳）
メタデータエラー:
行 |
  63 |  $statusCode = (New-Folder "D:\Temp")
     |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | "System.Object[]" 型の値 "System.Object[]" を "System.Int32" 型に変換することができません。
メイン処理が正常終了。statusCode[0]
```

前述している通り Function（`New-Folder`）を実行した戻り値が、System.Int32 ではなく 「**System.Object[]**」で返ってきたことがわかります。
では、実際にどのような流れで「**System.Object[]**」が返ってきたのかデバッグして詳細を確認してみましょう。

### 調査のためデバッグメッセージを追加したコード

Functionを実行する前後の `$statusCode` のデータ型と中身を確認します。
確認する方法として、`Write-Debug` を使いデバッグモード中にメッセージを出力するように修正。

PowerShellでデバッグモードをオンにするには、事前に下記を設定しておく必要があります。

```powershell:デバッグモードをオンにする設定
$DebugPreference = "Continue"
```

ちなみにデバッグモードから抜ける場合は下記を設定します。

```powershell:デバッグモードをオフにする設定（切り戻し用）
$DebugPreference = "SilentlyContinue"
```

デバッグメッセージは下記のように追加します。

```diff powershell:デバッグメッセージを追加
# メイン処理
[int]$statusCode = 0

+Write-Debug ($statusCode.GetType().FullName)
+Write-Debug "statusCode[$($statusCode)]"
$statusCode = (New-Folder "D:\Temp")
+Write-Debug ($statusCode.GetType().FullName)
+Write-Debug "statusCode[$($statusCode)]"

if ($statusCode -eq 0) {
    Write-Host "メイン処理が正常終了。statusCode[$($statusCode)]"
}
else {
    Write-Host "メイン処理が異常終了。statusCode[$($statusCode)]"
}
```

:::details 調査のためデバッグメッセージを追加したコード（全体）

デバッグ関連を追加したコード。
追加した箇所には「📑マーク」を入れています。

```powershell:
# 📑デバッグモードをオン
$DebugPreference = "Continue"
# $DebugPreference = "SilentlyContinue"

# Function群
function New-Folder {
    param (
        [string]$FolderPath
    )
    # 処理結果の進捗を入れる数値型のコード
    [int]$statusCode = 0

    if (Test-Path -Path $FolderPath) {
        # すでにフォルダーがある場合は削除
        try {
            Remove-Item -Path $FolderPath -Recurse -Force
        }
        catch {
            Write-Error "New-Folder: フォルダー削除でエラーが発生。FolderPath[$($FolderPath)]エラー[$($_.Exception.Message)]"
            $statusCode = -101
        }

        # 削除後にフォルダーを作成
        if ($statusCode -eq 0) {
            try {
                # ❌ New-Itemのコマンド結果が戻り値に入る
                New-Item -ItemType Directory -Path $FolderPath
                # ❌ Write-Outputのコマンド結果が戻り値に入る
                Write-Output "New-Folder: フォルダーを再作成。"
            }
            catch {
                Write-Error "New-Folder: フォルダー作成でエラーが発生。FolderPath[$($FolderPath)]エラー[$($_.Exception.Message)]"
                $statusCode = -201
            }
        }
    }
    else {
        # フォルダーがない場合はそのまま作成
        try {
            # ❌ New-Itemのコマンド結果が戻り値に入る
            New-Item -ItemType Directory -Path $FolderPath
            # ❌ Write-Outputのコマンド結果が戻り値に入る
            Write-Output "New-Folder: フォルダーを新規作成。"
        }
        catch {
            Write-Error "New-Folder: フォルダー作成でエラーが発生。FolderPath[$($FolderPath)]エラー[$($_.Exception.Message)]"
            $statusCode = -301
        }
    }

    # ❌ Write-Outputのコマンド結果が戻り値に入る
    if ($statusCode -eq 0) {
        Write-Output "New-Folderが正常終了。statusCode[$($statusCode)]"
    }
    else {
        Write-Output "New-Folderが異常終了。statusCode[$($statusCode)]"
    }

    # ❌ Intで返ったほしいが、標準出力の文字列型も入ってObject型として返してしまう。
    return $statusCode
}

# メイン処理
[int]$statusCode = 0

# 📑デバッグメッセージ追加
Write-Debug ($statusCode.GetType().FullName)
Write-Debug "statusCode[$($statusCode)]"
# ❌ Object型で戻ったことによりInt型に変換使用としてエラーが発生。$statusCodeは「0」のままとなる。
$statusCode = (New-Folder "D:\Temp")
# 📑デバッグメッセージ追加
Write-Debug ($statusCode.GetType().FullName)
Write-Debug "statusCode[$($statusCode)]"

# ❌ 上記でエラーが発生しているものの$statusCodeは「0」のままなので正常終了のメッセージが出力
if ($statusCode -eq 0) {
    Write-Host "メイン処理が正常終了。statusCode[$($statusCode)]"
}
else {
    Write-Host "メイン処理が異常終了。statusCode[$($statusCode)]"
}
```

:::

#### “調査のためデバッグメッセージを追加したコード”を実行した結果

```powershell:デバッグメッセージを追加したコードの実行結果
DEBUG: System.Int32
DEBUG: statusCode[0]
MetadataError:
Line |
  64 |  $statusCode = (New-Folder "D:\Temp")
     |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot convert the "System.Object[]" value of type "System.Object[]" to type "System.Int32".
DEBUG: System.Int32
DEBUG: statusCode[0]
メイン処理が正常終了。statusCode[0]
PS C:\Users\XXXX>
```

Function（`New-Folder`）を実行する前後で `$statusCode` の結果を確認すると、データ型は `System.Int32` で 中身は数字 `0` のままです。

つぎに `Cannot convert the "System.Object[]" value of type "System.Object[]" to type "System.Int32".` というメッセージを見ていきます。
Function（`New-Folder`）の戻り値は、PowerShell特有の仕様により `System.Object[]` で返りそのデータを数値型の `$statusCode` に代入しようとしているため、エラーが発生。
ここでのエラーは、データ型の変換を行おうとして発生したエラーとなります。これもPowerShellの仕様なのですが、別のデータ型に代入すると自動でキャスト（型の変換）が発生します。

そのキャスト処理が実行されるものの、`System.Object[]` から `System.Int32` はキャスト（変換）できないデータ型同士なので `Cannot convert` から始まる **変換エラーメッセージが発生した** という流れ。

もう少し具体的なデータの流れを説明すると本来、Function（`New-Folder`）で期待していたデータ型は、「`System.Int32`」。しかし、Function内の標準出力により「`@("New-Itemの標準出力結果", "Write-Outputの標準出力結果", "$statusCodeの数値")`」というような `System.Object[]` の戻り値が返ってきたということになります。

では、実際に戻ってきた `System.Object[]` の中身が予想通りか見てみましょう。

### 本当に戻り値が System.Object[] で標準出力も含まれているか確認

実際にFunction（`New-Folder`）の戻り値内を確認してみます。
確認するために新たにコードを追加！

#### 戻り値のオブジェクト型の中身をチェックするコード

```diff powershell:戻り値のオブジェクト型の中身をチェックするコードを追加
# メイン処理
-[int]$statusCode = 0
+[object]$resultData = @()

-$statusCode = (New-Folder "D:\Temp")
+$resultData = (New-Folder "D:\Temp")
+$resultData | Format-List

if ($statusCode -eq 0) {
    Write-Host "メイン処理が正常終了。statusCode[$($statusCode)]"
}
else {
    Write-Host "メイン処理が異常終了。statusCode[$($statusCode)]"
}
```

:::details 戻り値のオブジェクト型の中身をチェックするコード（全体）

上記のとおり、変数（`$resultData`）の中身を確認するためのコードを追加。
変化点には「👀マーク」を入れてコメントしてます。

```powershell:System.Object[]の中身確認用のコード
# Function群
function New-Folder {
    param (
        [string]$FolderPath
    )
    # 処理結果の進捗を入れる数値型のコード
    [int]$statusCode = 0

    if (Test-Path -Path $FolderPath) {
        # すでにフォルダーがある場合は削除
        try {
            Remove-Item -Path $FolderPath -Recurse -Force
        }
        catch {
            Write-Error "New-Folder: フォルダー削除でエラーが発生。FolderPath[$($FolderPath)]エラー[$($_.Exception.Message)]"
            $statusCode = -101
        }

        # 削除後にフォルダーを作成
        if ($statusCode -eq 0) {
            try {
                New-Item -ItemType Directory -Path $FolderPath
                Write-Output "New-Folder: フォルダーを再作成。"
            }
            catch {
                Write-Error "New-Folder: フォルダー作成でエラーが発生。FolderPath[$($FolderPath)]エラー[$($_.Exception.Message)]"
                $statusCode = -201
            }
        }
    }
    else {
        # フォルダーがない場合はそのまま作成
        try {
            New-Item -ItemType Directory -Path $FolderPath
            Write-Output "New-Folder: フォルダーを新規作成。"
        }
        catch {
            Write-Error "New-Folder: フォルダー作成でエラーが発生。FolderPath[$($FolderPath)]エラー[$($_.Exception.Message)]"
            $statusCode = -301
        }
    }
    
    if ($statusCode -eq 0) {
        Write-Output "New-Folderが正常終了。statusCode[$($statusCode)]"
    }
    else {
        Write-Output "New-Folderが異常終了。statusCode[$($statusCode)]"
    }
    
    return $statusCode
}

# メイン処理
# 👀 オブジェクト型として受け取るため変数を宣言
[object]$resultData = @()

# 👀 オブジェクト型として受け取る
$resultData = (New-Folder "D:\Temp")

# 👀 オブジェクト型の中身を確認
$resultData | Format-List

if ($statusCode -eq 0) {
    Write-Host "メイン処理が正常終了。statusCode[$($statusCode)]"
}
else {
    Write-Host "メイン処理が異常終了。statusCode[$($statusCode)]"
}
```

:::

#### “戻り値のオブジェクト型の中身をチェックするコード”を実行した結果

下記が実行した結果です。

```powershell:“戻り値のオブジェクト型の中身をチェックするコード”を実行した結果


    Directory: D:\

Name           : Temp
CreationTime   : 2025/03/13 12:28:42
LastWriteTime  : 2025/03/13 12:28:42
LastAccessTime : 2025/03/13 12:28:42
Mode           : d----
LinkType       :
Target         :

New-Folder: フォルダーを新規作成。
New-Folderが正常終了。statusCode[0]
0
メイン処理が正常終了。statusCode[0]
PS C:\Users\XXXX>
```

実行結果の一つひとつを分類してみると……

下記は「New-Itemコマンドレットの標準出力」の結果。

```powershell:New-Itemコマンドレットの標準出力

    Directory: D:\

Name           : Temp
CreationTime   : 2025/03/13 12:28:42
LastWriteTime  : 2025/03/13 12:28:42
LastAccessTime : 2025/03/13 12:28:42
Mode           : d----
LinkType       :
Target         :

```

つぎは「Function内の標準出力」の結果。

```powershell:Function内の標準出力
New-Folder: フォルダーを新規作成。
New-Folderが正常終了。statusCode[0]
```

その次は「Functionでreturnした数値型の戻り値」の結果。

```powershell:Functionでreturnした数値型の戻り値
0
```

さいごは「メイン処理側の標準出力」の結果。

```powershell:
メイン処理が正常終了。statusCode[0]
```

### 全体の流れを箇条書きで説明

ここまでのまとめとして箇条書きで処理の流れを記載します。

- メイン処理（Functionを呼び出す前）
    1. あらかじめステータスコードに「0」をセット
    1. Function（`New-Folder`）を呼び出し

- Function処理
    1. あらかじめステータスコードに「0」をセット
    1. 対象のフォルダーを作成
        **このタイミングで標準出力あり**
        `New-Item`コマンドレットにより標準出力され、明示的に指定が無くても戻り値として返されてしまう。
    1. Functionの処理結果を出力
        **このタイミングで標準出力あり**
        `Write-Output`コマンドレットにより標準出力され、明示的に指定が無くても戻り値として返されてしまう。
    1. ステータスコードを戻り値に返す
        **ココで戻り値がオブジェクト型に**
        これまでの処理で標準出力された結果が自動的に戻り値として返される。
        戻り値のオブジェクト型の中身は、`@("New-Itemの結果", "Write-Outputの結果", "ステータスコードの数値")` となる。

- メイン処理（Functionを呼び出した後）
    1. Functionの結果をステータスコードに代入
        **ココでエラーが発生**
        Functionの戻り値が **オブジェクト型** であり、**数値型** のステータスコードに代入しようと自動でデータ型の変換を行うとするが、
        **オブジェクト型** から **数値型** への変換は不可のため、変換エラーが発生。
        　
        なお、この代入時のエラーによりステータスコードは「0」のままで次へ
    1. メイン処理の終了メッセージ

## 解決方法

解決方法は簡単で「Function内では、**標準出力しないように変更**して想定通りの戻り値にする」だけです。

これまで題材にしたコードに対して 「**標準出力しないように変更**」してみます。

### 標準出力しないように変更したコード

`New-Item`などの標準出力を伴うコマンドレットは、`コマンド > $null` を使って標準出力を破棄します。

```diff powrshell
-New-Item -ItemType Directory -Path $FolderPath
+(New-Item -ItemType Directory -Path $FolderPath > $null)
```

`Write-Output` ではなく `Write-Host` により 標準出力 から コンソール出力 に変更。
※ コンソール出力とは、リダイレクトやパイプ処理が可能な標準出力はなく画面のみに出力する方式のこと。そのため、Function内で実行しても戻り値に影響しない。

```diff powrshell
-Write-Output "New-Folder: フォルダーを再作成。"
+Write-Host "New-Folder: フォルダーを再作成。"
```

:::detais 標準出力しないように変更したコード（全体）

変化点を「✅マーク」でコメントしています。

```powershell:標準出力しないように変更したコード
# Function群
function New-Folder {
    param (
        [string]$FolderPath
    )
    # 処理結果の進捗を入れる数値型のコード
    [int]$statusCode = 0

    if (Test-Path -Path $FolderPath) {
        # すでにフォルダーがある場合は削除
        try {
            Remove-Item -Path $FolderPath -Recurse -Force
        }
        catch {
            Write-Error "New-Folder: フォルダー削除でエラーが発生。FolderPath[$($FolderPath)]エラー[$($_.Exception.Message)]"
            $statusCode = -101
        }

        # 削除後にフォルダーを作成
        if ($statusCode -eq 0) {
            try {
                # ✅ New-Item 実行時の標準出力を > $null で破棄
                (New-Item -ItemType Directory -Path $FolderPath > $null)
                # ✅ Write-Output から Write-Host による変更で 標準出力 から コンソール（画面のみ） に！
                Write-Host "New-Folder: フォルダーを再作成。"
            }
            catch {
                Write-Error "New-Folder: フォルダー作成でエラーが発生。FolderPath[$($FolderPath)]エラー[$($_.Exception.Message)]"
                $statusCode = -201
            }
        }
    }
    else {
        # フォルダーがない場合はそのまま作成
        try {
            # ✅ New-Item 実行時の標準出力を > $null で破棄
            (New-Item -ItemType Directory -Path $FolderPath > $null)
            # ✅ Write-Output から Write-Host による変更で 標準出力 から コンソール（画面のみ） に！
            Write-Host "New-Folder: フォルダーを新規作成。"
        }
        catch {
            Write-Error "New-Folder: フォルダー作成でエラーが発生。FolderPath[$($FolderPath)]エラー[$($_.Exception.Message)]"
            $statusCode = -301
        }
    }

    # ✅ Write-Output から Write-Host による変更で 標準出力 から コンソール（画面のみ） に！
    if ($statusCode -eq 0) {
        Write-Host "New-Folderが正常終了。statusCode[$($statusCode)]"
    }
    else {
        Write-Host "New-Folderが異常終了。statusCode[$($statusCode)]"
    }

    # ✅ Function内で標準出力がないのでIntで返る
    return $statusCode
}

# メイン処理
[int]$statusCode = 0

# ✅ Functionの結果がIntで返り、変換することなくステータスコードに結果が代入
$statusCode = (New-Folder "D:\Temp")

# ✅ 実際の結果で分岐される
if ($statusCode -eq 0) {
    Write-Host "メイン処理が正常終了。statusCode[$($statusCode)]"
}
else {
    Write-Host "メイン処理が異常終了。statusCode[$($statusCode)]"
}
```

:::

## まとめ

最後に、もっとかみ砕いたわかりやすい例を元にまとめてみます。

### 事例「Function内で標準出力すると意図しない戻り値」

#### 問題

Function内で `Write-Output` のように標準出力を伴うコマンドレットを使用することで、意図せずその結果が戻り値として返される。

#### コード例

```powershell
function Get-Thirteen {
    Write-Output "Get-Thirteen 処理開始"    # ❌ この標準出力も戻り値として返る
    return 13
}

$result = Get-Thirteen
Write-Host "結果: $result"                  # "Get-Number 処理開始" と 13 の両方が返される
```

#### 対応方法の一覧

1. **標準出力が不要である場合、変更または削除する**
    標準出力が必須ではないのであれば、`Write-Host` などコンソール出力するコマンドレットに変更する。本当にいらない場合は `Write-Output` そのものを削除。
    `New-Item` をはじめとするコマンドレットの機能そのものは実行したいが標準出力は不要という場合は、下記のようにコマンドレットの標準出力を破棄する。
    - `コマンド > $null`    👈 Windows以外でも使用されていてパフォーマンスも速いらしい。
    - `コマンド | Out-Null` 👈 PowerShellのコマンドレットで全体的に統一感がでて可読性が良い。
    - `[void](コマンド)`    👈 可読性が悪く一般的にわかりにくい記述の仕方。
    - `$null = コマンド`    👈 同上。

    :::details コード例：標準出力が不要である場合、変更または削除する

    ```diff powershell
    function Get-Thirteen {
    -    Write-Output "Get-Thirteen 処理開始"    # ✅ 不要な標準出力を削除する
        return 13
    }

    $result = Get-Thirteen
    Write-Host "結果: $result"                  # 期待値の 13 が返される
    ```

    ```diff powershell
    function Get-Thirteen {
    -    Write-Output "Get-Thirteen 処理開始"
    +    Write-Host "Get-Thirteen 処理開始"      # ✅ コンソール出力に変更
        return 13
    }

    $result = Get-Thirteen
    Write-Host "結果: $result" -Debug            # 期待値の 13 が返される（コンソール出力あり）
    ```

    ```diff powershell
    function Get-Thirteen {
    -    Write-Output "Get-Thirteen 処理開始"
    +    Write-Output "Get-Thirteen 処理開始" > $null  # ✅ 標準出力を破棄する（Write-Outputの標準出力を破棄するのは変なコードですが……）
        return 13
    }

    $result = Get-Thirteen
    Write-Host "結果: $result"                  # 期待値の 13 が返される
    ```

    :::

1. **デバッグメッセージの場合はデバッグ用として設定する**
    調査フェーズのコード紹介で前述したとおり、デバッグメッセージを残す意図がある場合は `Write-Debug` を活用。
    実行時に「`-Debug`」というパラメーターを使用する方法でも簡単にデバッグモードで実行可能です。

    :::details コード例：デバッグメッセージの場合はデバッグ用として設定する

    ```diff powershell
    function Get-Thirteen {
    +    [CmdletBinding()]                       # ✅ Debugオプションを受付可能に
    +    param ()                                # ✅ 同上
    -    Write-Output "Get-Thirteen 処理開始"
    +    Write-Debug "Get-Thirteen 処理開始"      # ✅ デバッグ出力に変更
        return 13
    }

    -$result = Get-Thirteen
    +$result = (Get-Thirteen -Debug)              # ✅ デバッグモードでFunctionを実行
    Write-Host "結果: $result"                   # 期待値の 13 が返される（デバッグ出力あり）
    ```

    ```diff powershell
    function Get-Thirteen {
    -    Write-Output "Get-Thirteen 処理開始"
    +    Write-Debug "Get-Thirteen 処理開始"      # ✅ デバッグ出力に変更
        return 13
    }

    +$DebugPreference = "Continue"               # ✅ デバッグモードにする
    $result = Get-Thirteen
    Write-Host "結果: $result"                   # 期待値の 13 が返される（デバッグ出力あり）
    ```

    :::

1. **Function内の結果をどうしても標準出力したい場合は、戻り値を配列に変えよう**
    スクリプトの設計によってはFunction内で得た情報をどうしても標準出力したい場合があると思います。
    そのような場合は、あらかじめ設定する戻り値を配列にする方法がオススメです。
    配列が戻り値の場合のエラー制御ですが、最初に配列を初期化しておき配列にデータがあれば正常終了とし、
    データがない場合（`@()`などで判定）は異常終了するなどの動きが良いのでしょう。

    :::details コード例：Function内の結果をどうしても標準出力したい場合は、戻り値を配列に変えよう

    ```diff powershell
    function Get-Thirteen {
    +    [object[]]$objectArray = @()
    -    Write-Output "Get-Thirteen 処理開始"
        $objectArray += (Write-Output "Get-Thirteen 処理開始")
        $objectArray += 13
    -    return 13
    +    return $objectArray
    }

    $result = Get-Thirteen
    Write-Host "結果: $result"                  # "Get-Number 処理開始" と 13 の両方が返される
    ```

    :::

PowerShell独自

## 参考文献

https://zenn.dev/haretokidoki/articles/24ac3ba42d8050

https://stackoverflow.com/questions/5260125/whats-the-better-cleaner-way-to-ignore-output-in-powershell?form=MG0AV3&form=MG0AV3

## 関連記事
